File: rlm/plugins/discord.py
"""
Plugin Discord — rlm/plugins/discord.py

Integração com o Discord via Webhook URL e Bot API.

Arquitetura RLM-nativa:
    - Outbound via Webhook URL público (POST JSON, sem auth extra)
    - Outbound via Bot Token para canais específicos
    - DiscordAdapter registrado no ChannelRegistry como prefixo "discord"
    - O REPL chama diretamente: from rlm.plugins.discord import send_webhook

Variáveis de ambiente:
    DISCORD_WEBHOOK_URL   — URL padrão do webhook (canal default)
    DISCORD_BOT_TOKEN     — Token do bot (para enviar em canais por ID)
    DISCORD_APP_PUBLIC_KEY — Chave pública Ed25519 (para verificar interactions)

Uso no REPL:
    >>> from rlm.plugins.discord import send_webhook, send_embed
    >>> send_webhook("Deploy do backend concluído com sucesso!")
    '✓ enviado para discord'
    >>> send_embed("Relatório diário", "- 42 pedidos\\n- R$8.900 faturado", color=0x00FF00)
    '✓ embed enviado'
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from urllib import error as urllib_error
from urllib import request as urllib_request

try:
    from rlm.plugins import PluginManifest
except ImportError:
    from dataclasses import dataclass as _dc

    @_dc
    class PluginManifest:  # type: ignore
        name: str = ""
        version: str = ""
        description: str = ""
        functions: list = field(default_factory=list)
        author: str = ""
        requires: list = field(default_factory=list)


# ---------------------------------------------------------------------------
# Plugin Manifest
# ---------------------------------------------------------------------------

MANIFEST = PluginManifest(
    name="discord",
    version="1.0.0",
    description="Discord — envio via Webhook URL e Bot API (texto, embeds, arquivos).",
    functions=[
        "send_webhook",
        "send_embed",
        "send_channel_message",
        "pin_message",
        "create_thread",
        "get_channel_messages",
    ],
    author="RLM Engine",
    requires=[],  # stdlib urllib apenas
)


# ---------------------------------------------------------------------------
# Helpers internos
# ---------------------------------------------------------------------------

def _get_webhook_url() -> str:
    url = os.environ.get("DISCORD_WEBHOOK_URL", "")
    if not url:
        raise ValueError(
            "DISCORD_WEBHOOK_URL não configurada. "
            "Defina no .env: DISCORD_WEBHOOK_URL=https://discord.com/api/webhooks/..."
        )
    return url


def _get_bot_token() -> str:
    token = os.environ.get("DISCORD_BOT_TOKEN", "")
    if not token:
        raise ValueError(
            "DISCORD_BOT_TOKEN não configurado. "
            "Defina no .env: DISCORD_BOT_TOKEN=Bot seu_token"
        )
    # Discord exige o prefixo "Bot "
    if not token.startswith("Bot "):
        return f"Bot {token}"
    return token


def _discord_request(
    method: str,
    url: str,
    payload: dict | None = None,
    token: str | None = None,
) -> dict:
    """Realiza chamada HTTP à API do Discord (stdlib urllib)."""
    headers: dict[str, str] = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = token

    data: bytes | None = None
    if payload is not None:
        data = json.dumps(payload, ensure_ascii=False).encode("utf-8")

    req = urllib_request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib_request.urlopen(req, timeout=15) as resp:
            body = resp.read().decode("utf-8")
            return json.loads(body) if body else {"ok": True}
    except urllib_error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        return {"ok": False, "error": f"HTTP {e.code}: {body[:300]}"}
    except Exception as exc:
        return {"ok": False, "error": str(exc)}


# ---------------------------------------------------------------------------
# Funções REPL (chamadas diretamente pelo agente no REPL)
# ---------------------------------------------------------------------------

def send_webhook(
    text: str,
    webhook_url: str | None = None,
    username: str | None = None,
    avatar_url: str | None = None,
) -> str:
    """
    Envia mensagem de texto para um canal Discord via Webhook.

    O Webhook URL não precisa de autenticação — é suficiente para notificações.

    Args:
        text: Conteúdo da mensagem (até 2000 caracteres).
        webhook_url: URL do webhook. None = usa DISCORD_WEBHOOK_URL do ambiente.
        username: Sobrescreve o nome do bot exibido na mensagem.
        avatar_url: URL do avatar exibido (opcional).

    Returns:
        '✓ enviado para discord' em sucesso, string de erro em falha.
    """
    url = webhook_url or _get_webhook_url()
    payload: dict = {"content": text[:2000]}
    if username:
        payload["username"] = username
    if avatar_url:
        payload["avatar_url"] = avatar_url

    result = _discord_request("POST", url, payload)
    if result.get("ok") is False:
        return f"Erro Discord: {result.get('error', 'desconhecido')}"
    return "✓ enviado para discord"


def send_embed(
    title: str,
    description: str,
    color: int = 0x5865F2,
    webhook_url: str | None = None,
    fields: list[dict] | None = None,
    footer: str | None = None,
    url: str | None = None,
) -> str:
    """
    Envia um embed formatado para um canal Discord.

    Embeds suportam markdown, links, campos nomeados, rodapé e cor.

    Args:
        title: Título do embed (até 256 chars).
        description: Corpo do embed (até 4096 chars, suporta markdown).
        color: Cor da barra lateral em formato hexadecimal (ex: 0xFF0000 = vermelho).
        webhook_url: URL do webhook. None = usa DISCORD_WEBHOOK_URL do ambiente.
        fields: Lista de campos adicionais: [{"name": "Chave", "value": "Valor", "inline": True}].
        footer: Texto do rodapé.
        url: URL que o título linkará.

    Returns:
        '✓ embed enviado' em sucesso, string de erro em falha.
    """
    wh_url = webhook_url or _get_webhook_url()

    embed: dict = {
        "title": title[:256],
        "description": description[:4096],
        "color": color,
    }
    if url:
        embed["url"] = url
    if fields:
        embed["fields"] = [
            {
                "name": str(f.get("name", ""))[:256],
                "value": str(f.get("value", ""))[:1024],
                "inline": bool(f.get("inline", False)),
            }
            for f in fields[:25]  # limite Discord
        ]
    if footer:
        embed["footer"] = {"text": footer[:2048]}

    payload = {"embeds": [embed]}
    result = _discord_request("POST", wh_url, payload)
    if result.get("ok") is False:
        return f"Erro Discord embed: {result.get('error', 'desconhecido')}"
    return "✓ embed enviado"


def send_channel_message(channel_id: str, text: str, bot_token: str | None = None) -> str:
    """
    Envia mensagem para um canal Discord por ID usando Bot Token.

    Requer DISCORD_BOT_TOKEN no ambiente (ou bot_token explícito).
    O bot deve ter permissão "Send Messages" no canal alvo.

    Args:
        channel_id: ID numérico do canal Discord.
        text: Conteúdo da mensagem (até 2000 chars).
        bot_token: Token do bot. None = usa DISCORD_BOT_TOKEN do ambiente.

    Returns:
        ID da mensagem criada em sucesso, string de erro em falha.
    """
    token = bot_token or _get_bot_token()
    url = f"https://discord.com/api/v10/channels/{channel_id}/messages"
    payload = {"content": text[:2000]}

    result = _discord_request("POST", url, payload, token=token)
    if result.get("ok") is False or "id" not in result:
        err = result.get("error") or result.get("message", "desconhecido")
        return f"Erro Discord channel: {err}"
    return result["id"]


def pin_message(channel_id: str, message_id: str, bot_token: str | None = None) -> str:
    """
    Fixa uma mensagem em um canal Discord.

    Args:
        channel_id: ID do canal.
        message_id: ID da mensagem a fixar.
        bot_token: Token do bot. None = usa DISCORD_BOT_TOKEN do ambiente.

    Returns:
        '✓ mensagem fixada' ou string de erro.
    """
    token = bot_token or _get_bot_token()
    url = f"https://discord.com/api/v10/channels/{channel_id}/pins/{message_id}"
    result = _discord_request("PUT", url, token=token)
    if result.get("ok") is False:
        return f"Erro ao fixar: {result.get('error', 'desconhecido')}"
    return "✓ mensagem fixada"


def create_thread(
    channel_id: str,
    name: str,
    auto_archive_minutes: int = 1440,
    bot_token: str | None = None,
) -> str:
    """
    Cria um thread em um canal Discord.

    Args:
        channel_id: ID do canal pai.
        name: Nome do thread (até 100 chars).
        auto_archive_minutes: Inatividade para arquivamento (60, 1440, 4320, 10080).
        bot_token: Token do bot.

    Returns:
        ID do thread criado ou string de erro.
    """
    token = bot_token or _get_bot_token()
    url = f"https://discord.com/api/v10/channels/{channel_id}/threads"
    payload = {
        "name": name[:100],
        "auto_archive_duration": auto_archive_minutes,
        "type": 11,  # PUBLIC_THREAD
    }
    result = _discord_request("POST", url, payload, token=token)
    if result.get("ok") is False or "id" not in result:
        err = result.get("error") or result.get("message", "desconhecido")
        return f"Erro ao criar thread: {err}"
    return result["id"]


def get_channel_messages(
    channel_id: str,
    limit: int = 10,
    bot_token: str | None = None,
) -> list[dict]:
    """
    Retorna as últimas mensagens de um canal Discord.

    Args:
        channel_id: ID do canal.
        limit: Número de mensagens (1-100).
        bot_token: Token do bot.

    Returns:
        Lista de dicts com {id, author, content, timestamp} ou lista vazia em erro.
    """
    token = bot_token or _get_bot_token()
    n = max(1, min(100, limit))
    url = f"https://discord.com/api/v10/channels/{channel_id}/messages?limit={n}"
    result = _discord_request("GET", url, token=token)

    if isinstance(result, list):
        return [
            {
                "id": m.get("id"),
                "author": m.get("author", {}).get("username"),
                "content": m.get("content"),
                "timestamp": m.get("timestamp"),
            }
            for m in result
        ]
    return []


# ---------------------------------------------------------------------------
# Channel Registry Hook (DiscordAdapter)
# ---------------------------------------------------------------------------

try:
    from rlm.plugins.channel_registry import ChannelAdapter, ChannelRegistry

    class DiscordAdapter(ChannelAdapter):
        """
        Adapter Discord para o ChannelRegistry do RLM.

        Suporta dois modos por target_id:
        - "webhook" (ou vazio)   → posta no DISCORD_WEBHOOK_URL padrão
        - "<channel_id>"         → posta no canal via DISCORD_BOT_TOKEN

        Uso pelo REPL via ChannelRegistry:
            ChannelRegistry.reply("discord:webhook", "Deploy concluído")
            ChannelRegistry.reply("discord:1234567890", "Alerta: CPU alta")
        """

        def send_message(self, target_id: str, text: str) -> bool:
            """Envia texto ao target. 'webhook' = webhook URL, senão = channel_id."""
            if not target_id or target_id.lower() == "webhook":
                res = send_webhook(text)
            else:
                res = send_channel_message(target_id, text)
            return not res.startswith("Erro")

        def send_media(self, target_id: str, media_url_or_path: str, caption: str = "") -> bool:
            """
            Envia arquivo para um canal Discord.

            Para webhook: embeds com link para o arquivo.
            Para channel_id: mensagem com botão de link.
            """
            message = f"{caption}\n{media_url_or_path}".strip() if caption else media_url_or_path
            return self.send_message(target_id, message)

    # Auto-registro — ativado quando DISCORD_WEBHOOK_URL ou DISCORD_BOT_TOKEN estão presentes
    if os.environ.get("DISCORD_WEBHOOK_URL") or os.environ.get("DISCORD_BOT_TOKEN"):
        ChannelRegistry.register("discord", DiscordAdapter())

except ImportError:
    pass  # Permite uso standalone sem o registry

File: packages/gateway/src/adapters/discord.ts
/**
 * Discord Adapter (outbound) — envia respostas via Discord REST API
 *
 * Analogia Python: adapters/discord_adapter no rlm/plugins/
 *
 * Estratégia de envio:
 *   1. Se o routing_key contiver o interaction_id → usa follow-up de interaction
 *   2. Caso contrário → envia via webhook de canal (channel message)
 *
 * Variáveis de ambiente:
 *   RLM_DISCORD_BOT_TOKEN    — token do bot (Bot <token>)
 *   RLM_DISCORD_APP_ID       — application ID para follow-up de interactions
 */

import { childLogger } from "../logger.js";
import { chunkText } from "../chunker.js";
import type { ChannelAdapter, ChannelInfo, ProbeResult, SendResult } from "./interface.js";
import type { Envelope } from "../envelope.js";

const DISCORD_API = "https://discord.com/api/v10";
const DISCORD_MAX_CONTENT = 2_000;

export interface DiscordAdapterConfig {
  botToken: string;
  appId: string;
  timeoutMs?: number;
}

export class DiscordAdapter implements ChannelAdapter {
  readonly channelName = "discord";

  private readonly log = childLogger({ adapter: "discord" });
  private readonly config: Required<DiscordAdapterConfig>;

  private messagesSent = 0;
  private messagesReceived = 0;
  private errorCount = 0;
  private lastSeenMs: number | undefined;

  constructor(config: DiscordAdapterConfig) {
    this.config = { timeoutMs: 15_000, ...config };
  }

  async sendMessage(targetId: string, text: string, envelope: Envelope): Promise<SendResult> {
    this.lastSeenMs = Date.now();

    // targetId pode ser "channelId" ou "interactionId:token"
    const interactionToken = (envelope.metadata["discord_interaction_token"] as string | undefined);
    const interactionId = (envelope.metadata["discord_interaction_id"] as string | undefined);

    const chunks = chunkText(text, { maxLength: DISCORD_MAX_CONTENT });
    let lastResult: SendResult = { ok: false };

    for (let i = 0; i < chunks.length; i++) {
      const chunk = chunks[i]!;

      if (interactionId && interactionToken && i === 0) {
        lastResult = await this.sendInteractionFollowup(interactionId, interactionToken, chunk);
      } else if (interactionId && interactionToken && i > 0) {
        lastResult = await this.sendInteractionFollowup(interactionId, interactionToken, chunk);
      } else {
        lastResult = await this.sendChannelMessage(targetId, chunk);
      }

      if (!lastResult.ok) break;
    }

    return lastResult;
  }

  async sendMedia(targetId: string, url: string, _mime: string, caption?: string): Promise<SendResult> {
    this.lastSeenMs = Date.now();
    // Discord suporta embeds com URL imagem
    const payload: Record<string, unknown> = {
      embeds: [{ image: { url }, description: caption ?? "" }],
    };
    return this.apiPost(`/channels/${targetId}/messages`, payload);
  }

  getChannelInfo(): ChannelInfo {
    return {
      id: "discord",
      name: "Discord",
      type: "bot",
      status: this.config.botToken ? "healthy" : "disabled",
      ...(this.lastSeenMs !== undefined ? { lastSeenMs: this.lastSeenMs } : {}),
      messagesSent: this.messagesSent,
      messagesReceived: this.messagesReceived,
      errors: this.errorCount,
    };
  }

  async probe(timeoutMs = this.config.timeoutMs): Promise<ProbeResult> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    const startedAt = Date.now();

    try {
      const response = await fetch(`${DISCORD_API}/users/@me`, {
        method: "GET",
        headers: {
          "Authorization": `Bot ${this.config.botToken}`,
          "User-Agent": "RLM-Gateway/1.0",
        },
        signal: controller.signal,
      });

      if (!response.ok) {
        const errText = await response.text();
        this.errorCount++;
        return {
          ok: false,
          elapsedMs: Date.now() - startedAt,
          error: `HTTP ${response.status}: ${errText}`,
        };
      }

      const data = await response.json() as {
        id?: string;
        username?: string;
        global_name?: string | null;
      };

      this.lastSeenMs = Date.now();
      return {
        ok: true,
        elapsedMs: Date.now() - startedAt,
        identity: {
          ...(data.id !== undefined ? { botId: data.id } : {}),
          ...(data.username !== undefined ? { username: data.username } : {}),
          ...((data.global_name ?? data.username) !== undefined ? { displayName: data.global_name ?? data.username } : {}),
        },
      };
    } catch (err) {
      this.errorCount++;
      return {
        ok: false,
        elapsedMs: Date.now() - startedAt,
        error: String(err),
      };
    } finally {
      clearTimeout(timer);
    }
  }

  // --------------------------------------------------------------------------
  // Discord REST helpers
  // --------------------------------------------------------------------------

  private async sendChannelMessage(channelId: string, content: string): Promise<SendResult> {
    return this.apiPost(`/channels/${channelId}/messages`, { content });
  }

  private async sendInteractionFollowup(
    appId: string,
    interactionToken: string,
    content: string,
  ): Promise<SendResult> {
    return this.apiPost(`/webhooks/${appId}/${interactionToken}`, { content });
  }

  private async apiPost(path: string, body: Record<string, unknown>): Promise<SendResult> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.config.timeoutMs);

    try {
      const res = await fetch(`${DISCORD_API}${path}`, {
        method: "POST",
        headers: {
          "Authorization": `Bot ${this.config.botToken}`,
          "Content-Type": "application/json",
          "User-Agent": "RLM-Gateway/1.0",
        },
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      if (!res.ok) {
        const errText = await res.text();
        this.errorCount++;
        this.log.error({ status: res.status, body: errText, path }, "Discord API error");
        return { ok: false, error: `HTTP ${res.status}: ${errText}` };
      }

      const data = await res.json() as { id?: string };
      this.messagesSent++;
      return { ok: true, ...(data.id !== undefined ? { messageId: data.id } : {}) };
    } catch (err) {
      this.errorCount++;
      this.log.error({ err, path }, "Discord fetch error");
      return { ok: false, error: String(err) };
    } finally {
      clearTimeout(timeout);
    }
  }

  static fromEnv(): DiscordAdapter | null {
    const token = process.env["RLM_DISCORD_BOT_TOKEN"];
    const appId = process.env["RLM_DISCORD_APP_ID"];
    if (!token || !appId) return null;
    return new DiscordAdapter({ botToken: token, appId });
  }
}
